enso-org/enso
# Lift asset objects to global cache

### Pull Request Description
- Lift state of individual assets all the way to the global React Query cache
  - Previously, the state of assets were contained only in the individual rows representing assets.

### Important Notes
- There are some follow-up refactors that will be enabled by this PR:
  - Refactor the Data Catalog to have a nested component structure, rather than the current flat structure that requires a custom object to handle e.g. sorting
  - Switching the nested components to a React Query query to fetch the list of children. This means:
    - We can migrate all placeholder rows to a query wrapper by temporarily applying in-flight mutations to the query result.
    - We will also avoid the complex fetching logic
    - Directory state can/will be refreshed individually.
    - Importantly - `specialLoading` and `specialEmpty` special placeholder row types will no longer be necessary, as the nested component has the query's loading state, *and* knows whether the list of its children has length 0 (is empty).
  - Directories should not maintain a list of children - only a list of child IDs
  - `setItem`/`AssetTreeNode`/`nodeMapRef` will be made redundant, greatly simplifying the architecture.

### Testing instructions
- *All* actions that modify assets table state should be tested:
  - Uploading new files with the same name
  - Opening a project
  - Renaming an open project from the Data Catalog (and checking that the name is changed in the Editor)
  - Renaming an open project from the Editor (and checking that the name is changed in the Data Catalog)
  - Creating every type of asset
  - Copying/pasting/cutting/moving assets to a different directory and back

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
